### PR TITLE
fixes #15916 - infrastructure to switch tests to use vcr 

### PR DESCRIPTION
**Pull request by thomasmckay**

### CODE REVIEW

**Comment by thomasmckay** [comment]
Ran w/
```
rake test mode=all record=false test=resources/settings
```

**Comment by thomasmckay** [comment]
```
$ rake test logging=true mode=all record=true test=resources/settings
Debugging not enabled.
Running tests for: resources/settings
Could not load the API description from the server
  - is the server down?
  - was 'foreman-rake apipie:cache' run on the server when using apipie cache? (typical production settings)
rake aborted!
Could not load data from http://katello:3000
 - is your server down?
 - was rake apipie:cache run when using apipie cache? (typical production settings)
/home/thomasmckay/code/hammer-cli-foreman/lib/hammer_cli_foreman.rb:144:in `rescue in <module:HammerCLIForeman>'
/home/thomasmckay/code/hammer-cli-foreman/lib/hammer_cli_foreman.rb:25:in `<module:HammerCLIForeman>'
/home/thomasmckay/code/hammer-cli-foreman/lib/hammer_cli_foreman.rb:5:in `<top (required)>'
/home/thomasmckay/code/hammer-cli-csv/lib/hammer_cli_csv.rb:5:in `require'
/home/thomasmckay/code/hammer-cli-csv/lib/hammer_cli_csv.rb:5:in `<module:HammerCLICsv>'
/home/thomasmckay/code/hammer-cli-csv/lib/hammer_cli_csv.rb:4:in `<top (required)>'
/home/thomasmckay/code/hammer-cli-csv/test/csv_test_helper.rb:26:in `require'
/home/thomasmckay/code/hammer-cli-csv/test/csv_test_helper.rb:26:in `<top (required)>'
/home/thomasmckay/code/hammer-cli-csv/test/resources/settings_test.rb:4:in `require'
/home/thomasmckay/code/hammer-cli-csv/test/resources/settings_test.rb:4:in `<top (required)>'
/home/thomasmckay/code/hammer-cli-csv/test/test_runner.rb:163:in `require'
/home/thomasmckay/code/hammer-cli-csv/test/test_runner.rb:163:in `run_tests'
/home/thomasmckay/code/hammer-cli-csv/Rakefile:43:in `block (3 levels) in <top (required)>'
/home/thomasmckay/code/hammer-cli-csv/Rakefile:77:in `block in <top (required)>'
/home/thomasmckay/.rvm/gems/ruby-2.0.0-p481@hammer/bin/ruby_executable_hooks:15:in `eval'
/home/thomasmckay/.rvm/gems/ruby-2.0.0-p481@hammer/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => test:resources
(See full trace by running task with --trace)
Coverage report generated for MiniTest to /home/thomasmckay/code/coverage. 1319 / 3234 LOC (40.79%) covered.
Run options: --seed 11559

# Running tests:

Running Suite MiniTest::Spec - :test 
Completed Running Suite MiniTest::Spec - :test 


Finished tests in 0.000318s, 0.0000 tests/s, 0.0000 assertions/s.

0 tests, 0 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by akofink** [comment]
Was there an rfc for this? I totally support adding tests, but let's discuss the plan. hammer-cli-katello doesn't use VCR, simply the apipie cache as its input. Could that be an option? Are there already bats tests? I can't seem to find any in https://github.com/Katello/forklift/tree/master/bats. Essentially, I'd like to see some discussion on why we think VCR is the way forward and also how this test suite can be maintainable with an app API that could potentially change.

**Comment by thomasmckay** [comment]
I'd like to see bats tests as well, yes. This would catch the myriad issues I've had w/ the API changing.

For VCR, I'd like tests that confirm that export/import remains consistent, especially since there is going to be some logic added around attaching subscriptions to hosts. Running in live mode is equivalent to the tests I have now.

What specifically do you see as potential issues in adding VCR? I'm happy to adjust the strategy. In the end, I want live tests as well as non-live.

**Comment by akofink** [comment]
I'm not really familiar with VCR nor have any qualms about it; just wondering why we've chosen it. Do we already use it in some other repository?

Doing a bit of research on web mocking frameworks, there are [many to choose from](https://www.ruby-toolbox.com/categories/mocking_web_requests). It seems like webmock is lighter and more maintainable, according to [this blog post from 2015](https://quickleft.com/blog/decreasing-vcr-dependency-webmock/), but it requires a bit more work up front (Yes, it's only one opinion/experience).

I do find it strange that a test framework would have [such mediocre coverage](https://codeclimate.com/github/vcr/vcr/coverage?sort=covered_percent&sort_direction=asc). I'm playing a bit of devil's advocate if you can't tell. :smiling_imp: As we discussed via IRC last week, we should totally pick the easiest way to get tests in hammer-cli-csv first so we get the most bang for the buck, so to speak, and ideally add more fine-grained or maintainable tests over time. If VCR will meet that requirement, so be it. :+1:

**Comment by thomasmckay** [comment]
@akofink - Definitely bring discussion to my thread on foreman-dev, if desired. VCR is used in our repos; the example I was pointed to as a good example is https://github.com/Katello/runcible

**Comment by daviddavis** [comment]
@akofink I'm not sure I would call VCR a web mocking framework. It's more of a tool that allows you to record web requests and then we replay those using webmock. We use VCR quite extensively in Katello/katello and I haven't heard of any problems. Our requests are complex so being able to record actual requests and use them as mocks is much easier than manually constructing them in the test code.

**Comment by thomasmckay** [comment]
[test]

**Comment by thomasmckay** [comment]
@akofink @daviddavis all set for ack/nack

**Comment by thomasmckay** [comment]
[test]

File: README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# hammer-cli-csv ![Travis # Status](https://travis-ci.org/Katello/hammer-cli-csv.svg?branch=master)
+# hammer-cli-csv ![Travis # Status](https://travis-ci.org/Katello/hammer-cli-csv.svg?branch=master) [![Coverage Status](https://coveralls.io/repos/github/Katello/hammer-cli-csv/badge.svg?branch=master)](https://coveralls.io/github/Katello/hammer-cli-csv?branch=master)
```
Comment (akofink) [inline README.md]: This badge is wrong. It should be `[Build Status](https://travis-ci.org/Katello/hammer-cli-csv)`

File: README.md [hunk 2]
```diff
@@ -556,11 +556,31 @@ Updating setting 'idle_timeout'...done
...
+```record=false```
+
+To see RestClient logs while testing:
+
+```logging=true```
+
+Test server configuration is taken from 'test/config.yml'. If that file does not exist then https://localhost with admin / changeme is the default.
 
 ```
 % cat test/config.yml
```
Comment (akofink) [inline README.md]: Should this be in a ``` ?

File: README.md (hunk 1 above)
Comment (thomasmckay) [inline README.md]: You sure? https://raw.githubusercontent.com/Katello/hammer-cli-katello/master/README.md

Comment (thomasmckay) [inline README.md]: Oh, I see. I'll change.

File: README.md [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# hammer-cli-csv ![Travis # Status](https://travis-ci.org/Katello/hammer-cli-csv.svg?branch=master) [![Coverage Status](https://coveralls.io/repos/github/Katello/hammer-cli-csv/badge.svg?branch=master)](https://coveralls.io/github/Katello/hammer-cli-csv?branch=master)
+# hammer-cli-csv [![Build Status](https://travis-ci.org/Katello/hammer-cli-csv.svg?branch=master)(https://travis-ci.org/Katello/hammer-cli-csv)] [![Coverage Status](https://coveralls.io/repos/github/Katello/hammer-cli-csv/badge.svg?branch=master)](https://coveralls.io/github/Katello/hammer-cli-csv?branch=master)
```
Comment (akofink) [inline README.md]: Hmm... not quite. Maybe it's because it's inside the `<h1>` (#) ?

**Comment by akofink** [comment]
I believe you need to add a default rake task. It should probably run rubocop, then the tests.

File: README.md (hunk 1 above)
Comment (akofink) [inline README.md]: Like this:

[Build Status](https://travis-ci.org/Katello/hammer-cli-csv)

**Comment by akofink** [comment]
APJ thanks @thomasmckay!

**Pull request merged**